### PR TITLE
Fix getting the window owner name on Windows

### LINKED ISSUE #130: activeWindow returns a process name instead of a product name to windows owner.name
In Windows, when getting information about the active window, the process name is returned instead of the "normal" product name

For example, for the Chrome browser it returns
owner.name = "crome.exe" instead of "Google Chrome"

### PR DESCRIPTION
**Pull request by pityulin**
Fixes #130

### CODE REVIEW

**Comment by sindresorhus** [comment]
Linting is failing.

**Comment by sindresorhus** [comment]
have you thoroughly tested your changes? You're changing a lot of stuff and I don't have Windows to test at the moment.

File: lib/windows.js [hunk 1]
```diff
@@ -72,9 +72,52 @@ const kernel32 = new ffi.Library('kernel32', {
...
+
+function getVarEntry(mem) {
+	const val = ref.alloc(VoidPointer);
+	const size = ref.alloc('uint32', 1);
+	const res = version.VerQueryValueW(mem, wstring(String.raw`\\VarFileInfo\\Translation`), val, size);
+	if (res) {
+		const num = ref.readPointer(val, 0, 4);
+		return num.readUInt16LE(0) << 16 | num.readUInt16LE(2);
+	}
+	return 0x040904E4;
```
Comment (sindresorhus) [inline lib/windows.js:95]: This should be a constant with a descriptive name.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
I'd like to see some more code comments.

File: lib/windows.js [hunk 2]
```diff
@@ -72,9 +72,54 @@ const kernel32 = new ffi.Library('kernel32', {
...
+	GetFileVersionInfoW: ['int32', ['string', 'uint32', 'uint32', 'pointer']],
+	VerLanguageNameW: ['int32', ['uint32', 'pointer', 'uint32']],
+	VerQueryValueW: ['int32', ['pointer', 'pointer', VoidPointer, 'uint32 *']]
+});
+
+function wstring(string) {
+	return ref.allocCString(string + '\0', 'ucs2');
+}
+
+function getLangId(mem) {
```
Comment (sindresorhus) [inline lib/windows.js:87]:
```suggestion
function getLanguageId(mem) {
```

File: lib/windows.js [hunk 3]
```diff
@@ -72,9 +72,54 @@ const kernel32 = new ffi.Library('kernel32', {
...
+// Create FFI declarations for the C++ library and functions needed (version.dll)
+const version = new ffi.Library('version', {
+	GetFileVersionInfoSizeW: ['uint32', ['string', 'uint32']],
+	GetFileVersionInfoW: ['int32', ['string', 'uint32', 'uint32', 'pointer']],
+	VerLanguageNameW: ['int32', ['uint32', 'pointer', 'uint32']],
+	VerQueryValueW: ['int32', ['pointer', 'pointer', VoidPointer, 'uint32 *']]
+});
+
+function wstring(string) {
+	return ref.allocCString(string + '\0', 'ucs2');
```
Comment (sindresorhus) [inline lib/windows.js:84]:
```suggestion
	return ref.allocCString(`${string}\0`, 'ucs2');
```

File: lib/windows.js [hunk 4]
```diff
@@ -72,9 +72,54 @@ const kernel32 = new ffi.Library('kernel32', {
...
 
+// Create FFI declarations for the C++ library and functions needed (version.dll)
+const version = new ffi.Library('version', {
+	GetFileVersionInfoSizeW: ['uint32', ['string', 'uint32']],
+	GetFileVersionInfoW: ['int32', ['string', 'uint32', 'uint32', 'pointer']],
+	VerLanguageNameW: ['int32', ['uint32', 'pointer', 'uint32']],
+	VerQueryValueW: ['int32', ['pointer', 'pointer', VoidPointer, 'uint32 *']]
+});
+
+function wstring(string) {
```
Comment (sindresorhus) [inline lib/windows.js:83]:
```suggestion
function toWstring(string) {
```

File: lib/windows.js [hunk 5]
```diff
@@ -72,9 +72,54 @@ const kernel32 = new ffi.Library('kernel32', {
...
+	}
+
+	return 0x040904E4;
+}
+
+function hexString(value) {
+	return ('00000000' + value.toString(16)).slice(-8);
+}
+
+function getFileVersionString(mem, name) {
```
Comment (sindresorhus) [inline lib/windows.js:103]: > mem

Don't use abbreviations

File: lib/windows.js [hunk 6]
```diff
@@ -72,9 +72,54 @@ const kernel32 = new ffi.Library('kernel32', {
...
+	const value = ref.alloc(VoidPointer);
+	const size = ref.alloc('uint32', 0);
+	const result = version.VerQueryValueW(mem, wstring(name), value, size);
+	let data;
+	if (result) {
+		const buf = ref.readPointer(value, 0, size.deref() * wchar.size);
```
Comment (sindresorhus) [inline lib/windows.js:109]:
```suggestion
		const buffer = ref.readPointer(value, 0, size.deref() * wchar.size);
```

File: .vscode/launch.json [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+            "type": "pwa-node",
+            "request": "launch",
+            "name": "Launch Program",
+            "skipFiles": [
+                "<node_internals>/**"
+            ],
+            "program": "${workspaceFolder}\\test1.js"
+        }
+    ]
+}
```
Comment (sindresorhus) [inline .vscode/launch.json:17]: This should not be committed.

**Comment by sindresorhus** [comment]
> have you thoroughly tested your changes?

I never got a reply to this.

**Comment by sindresorhus** [comment]
Linting is failing.

File: lib/windows.js [hunk 8]
```diff
@@ -123,8 +165,22 @@ function getProcessPath(processHandle) {
...
+	const unused = ref.alloc('uint32', 1);
+	const infoSize = version.GetFileVersionInfoSizeW(processFileNameBufferClean, unused);
+	if (infoSize !== 0) {
+		// Get data from the FileVersionInfo struct
+		const block = new Buffer.alloc(infoSize);
+		if (version.GetFileVersionInfoW(processFileNameBufferClean, 0, infoSize, block)) {
+			// Get file description from the FileVersionInfo struct
+			const langId = getLanguageId(block);
+			const fileDescription = getVersionInfoForCodePage(block, hexString(langId));
+			processName = fileDescription.trim() || processName;
```
Comment (supercoww) [inline lib/windows.js:179]: `fileDescription` can be undefined for some apps. For me it crashed when opening Windows Terminal or Snipping tool.

```
D:\projects\active-win\node_modules\.pnpm\ffi-napi@4.0.3\node_modules\ffi-napi\lib\_foreign_function.js:61      
    bindings.ffi_call(cif, funcPtr, result, argsList);
             ^

Error: TypeError: Cannot read properties of undefined (reading 'trim')
    at Object.proxy [as EnumChildWindows] (D:\projects\active-win\node_modules\.pnpm\ffi-napi@4.0.3\node_modules\ffi-napi\lib\_foreign_function.js:61:14)    at getSubWindowRealProcessNameAndPath (D:\projects\active-win\lib\windows.js:143:9)
    at Function.windows [as sync] (D:\projects\active-win\lib\windows.js:236:32)
    at Function.module.exports.sync (D:\projects\active-win\index.js:29:38)
```

**Pull request merged**